linfa-reduction/src/fast_ica/mod.rs

```diff
@@ -0,0 +1,382 @@
+    pub fn fit<A: Float>(&self, x: &Array2<A>) -> Result<FittedFastIca<A>, String> {
+        let (nsamples, nfeatures) = (x.nrows(), x.ncols());
+
+        // If the number of components is not set, we take the minimum of
+        // the number of rows and columns
+        let ncomponents = self.ncomponents.unwrap_or_else(|| nsamples.min(nfeatures));
+
+        // The number of components cannot be greater than the minimum of
+        // the number of rows and columns
+        if ncomponents > nsamples.min(nfeatures) {
+            return Err(
+                "ncomponents cannot be greater than the min(sample size, features size)"
+                    .to_string(),
+            );
+        }
+
+        // We center the input by subtracting the mean of its features
+        let xmean = x.mean_axis(Axis(0)).unwrap();
+        let mut xcentered = x - &xmean.to_owned().insert_axis(Axis(0));
+
+        // We transpose the centered matrix
+        xcentered = xcentered.reversed_axes();
+
+        // We whiten the matrix to remove any potential correlation between
+        // the components
+        let k = match xcentered.svd(true, false).unwrap() {
+            (Some(u), s, _) => {
+                let s = s.mapv(|x| A::from(x).unwrap());
+                (u.slice(s![.., ..nsamples.min(nfeatures)]).to_owned() / s)
+                    .t()
+                    .slice(s![..ncomponents, ..])
+                    .to_owned()
+            }
+            _ => unreachable!(),
+        };
+        let mut xwhitened = k.dot(&xcentered);
+
+        // We multiply the matrix with root of the number of records
+        let nsamples_sqrt = A::from((nsamples as f64).sqrt()).unwrap();
+        xwhitened.mapv_inplace(|x| x * nsamples_sqrt);
+
+        // We initialize the de-mixing matrix with a uniform distributiona
+        let w: Array2<f64>;
+        if let Some(seed) = self.random_state {
+            let mut rng = Isaac64Rng::seed_from_u64(seed as u64);
+            w = Array::random_using((ncomponents, ncomponents), Uniform::new(0., 1.), &mut rng);
+        } else {
+            w = Array::random((ncomponents, ncomponents), Uniform::new(0., 1.));
+        }
+        let mut w = w.mapv(|x| A::from(x).unwrap());
+
+        // We find the optimized de-mixing matrix
+        w = self.ica_parallel(&xwhitened, &w)?;
+
+        // We whiten the de-mixing matrix
+        let components = w.dot(&k);
+
+        Ok(FittedFastIca {
+            mean: xmean,
+            components,
+        })
+    }
+
+    // Parallel FastICA, Optimization step
+    fn ica_parallel<A: Float>(&self, x: &Array2<A>, w: &Array2<A>) -> Result<Array2<A>, String> {
+        let mut w = Self::sym_decorrelation(&w);
+
+        let p = x.ncols() as f64;
+
+        for _ in 0..self.max_iter {
+            let (gwtx, g_wtx) = self.gfunc.exec(&w.dot(x))?;
+
+            let lhs = gwtx.dot(&x.t()).mapv(|x| x / A::from(p).unwrap());
+            let rhs = &w * &g_wtx.insert_axis(Axis(1));
```

Comment:
isn't a `1/N` missing here, because you want to calculate the mean value of the gradient?

Reply:
`p` here is N, x is transposed so `p` which is `x.ncols()` would be N